### PR TITLE
Added options for hvd.DistributedOptimzer

### PR DESCRIPTION
**Pull request by pescap**
Tested! Working as expected!

### CODE REVIEW

**Comment by pescap** [comment]
With respect to the Codacy alert, I am not sure to understand why I get it for `hvd_opt_options`  while `LBFGS_options` is fine.

File: deepxde/optimizers/config.py [hunk 1]
```diff
@@ -1,9 +1,16 @@
...
+from ..config import hvd
 
+compression = None
+op = None
+if hvd is not None:
+    compression = hvd.compression.Compression.none
+    op = hvd.Average
 
 LBFGS_options = {}
+hvd_opt_options = {}
```
Comment (lululxvi) [inline deepxde/optimizers/config.py:13]:
```
if hvd is not None:
    hvd_opt_options = {}
    compression = hvd.compression.Compression.none
    op = hvd.Average
```

File: deepxde/optimizers/config.py [hunk 2]
```diff
@@ -59,7 +66,48 @@ def set_LBFGS_options(
     LBFGS_options["maxls"] = maxls
 
 
+def set_hvd_opt_options(
+    compression=compression,
+    op=op,
```
Comment (lululxvi) [inline deepxde/optimizers/config.py:71]:
```
compression=hvd.compression.Compression.none,
op=hvd.Average,
```

File: deepxde/optimizers/config.py [hunk 3]
```diff
@@ -1,9 +1,16 @@
...
+__all__ = ["set_LBFGS_options", "set_hvd_opt_options"]
 
 from ..backend import backend_name
+from ..config import hvd
 
+compression = None
+op = None
+if hvd is not None:
+    compression = hvd.compression.Compression.none
+    op = hvd.Average
```
Comment (lululxvi) [inline deepxde/optimizers/config.py:10]: These 5 lines do not seem useful.

File: deepxde/optimizers/tensorflow_compat_v1/optimizers.py [hunk 4]
```diff
@@ -56,7 +56,17 @@ def get(loss, optimizer, learning_rate=None, decay=None):
...
+            "optimizer": optim,
+            "compression": hvd_opt_options["compression"],
+            "op": hvd_opt_options["op"],
+            "backward_passes_per_step": hvd_opt_options["backward_passes_per_step"],
+            "average_aggregated_gradients": hvd_opt_options[
+                "average_aggregated_gradients"
+            ],
+        }
+
+        optim = hvd.DistributedOptimizer(**options)
```
Comment (lululxvi) [inline deepxde/optimizers/tensorflow_compat_v1/optimizers.py:69]:
```
optim = hvd.DistributedOptimizer(
    optim,
    compression=hvd_opt_options["compression"],
    op=hvd_opt_options["op"],
    backward_passes_per_step=hvd_opt_options["backward_passes_per_step"],
    average_aggregated_gradients=hvd_opt_options["average_aggregated_gradients"],
)
```

File: deepxde/optimizers/config.py (hunk 3 above)
Comment (pescap) [reply deepxde/optimizers/config.py]: I defined it like this as `compression=hvd.compression.Compression.none` gives an error if `hvd is None` (refer to the last commit).

Shall I revert the changes?

Comment (pescap) [reply deepxde/optimizers/config.py]: I defined it like this as compression=hvd.compression.Compression.none gives an error if hvd is None (refer to my last commit and the errors for the workflow).

Shall I revert the changes?

File: deepxde/optimizers/tensorflow_compat_v1/optimizers.py (hunk 4 above)
Comment (pescap) [reply deepxde/optimizers/tensorflow_compat_v1/optimizers.py]: See comment above.

File: deepxde/optimizers/config.py [hunk 5]
```diff
@@ -59,7 +63,48 @@ def set_LBFGS_options(
     LBFGS_options["maxls"] = maxls
 
 
+def set_hvd_opt_options(
+    compression=hvd.compression.Compression.none,
+    op=hvd.Average,
```
Comment (lululxvi) [inline deepxde/optimizers/config.py:68]:
```
compression=None,
op=None,
```
Then in the code, if `None`, use the default values.

File: deepxde/optimizers/config.py [hunk 6]
```diff
@@ -1,9 +1,13 @@
-__all__ = ["set_LBFGS_options"]
+__all__ = ["set_LBFGS_options", "set_hvd_opt_options"]
 
 from ..backend import backend_name
+from ..config import hvd
 
-
+compression = None
+op = None
```
Comment (lululxvi) [inline deepxde/optimizers/config.py:7]: Delete lines 6 and 7.

**Comment by pescap** [comment]
> With respect to the Codacy alert, I am not sure to understand why I get it for `hvd_opt_options` while `LBFGS_options` is fine.

Well, it works fine without the use of `global`.

**Comment by lululxvi** [comment]
> Well, it works fine without the use of `global`.

Really? Without `global`, you can also change `hvd_opt_options`?

File: deepxde/optimizers/__init__.py [hunk 7]
```diff
@@ -1,7 +1,7 @@
 import importlib
 import sys
 
-from .config import LBFGS_options, set_LBFGS_options
+from .config import LBFGS_options
```
Comment (lululxvi) [inline deepxde/optimizers/__init__.py:4]: Why delete `set_LBFGS_options`?

File: deepxde/optimizers/config.py [hunk 8]
```diff
@@ -59,7 +61,51 @@ def set_LBFGS_options(
...
+def set_hvd_opt_options(
+    compression=None,
+    op=None,
+    backward_passes_per_step=1,
+    average_aggregated_gradients=False,
+):
+    """Sets the parameters of hvd.DistributedOptimizer.
+
+    The default parameters are the same as for hvd.DistributedOptimizer.
+    Refer to `hvd.DistributedOptimizer <https://horovod.readthedocs.io/en/stable/api.html>`_
```
Comment (lululxvi) [inline deepxde/optimizers/config.py:73]: The default parameters are the same as for `hvd.DistributedOptimizer <https://horovod.readthedocs.io/en/stable/api.html>`_.

File: deepxde/optimizers/config.py [hunk 9]
```diff
@@ -59,7 +61,51 @@ def set_LBFGS_options(
...
+
+    Args:
+        compression: Compression algorithm used to reduce the amount of data
+                     sent and received by each worker node.  Defaults to not
+                     using compression.
```
Comment (lululxvi) [inline deepxde/optimizers/config.py:78]: Use only 4 spaces for indent.

File: deepxde/optimizers/config.py [hunk 10]
```diff
@@ -59,7 +61,51 @@ def set_LBFGS_options(
...
+                                      have been accumulated over multiple mini-batches.
+                                      If true divides gradient updates by
+                                      backward_passes_per_step.
+
+    Warning:
+        hvd is only implemented for tensorflow.compat.v1.
+
+        .. code-block:: python
+
+            dde.config.set_hvd_opt_options()
```
Comment (lululxvi) [inline deepxde/optimizers/config.py:94]: Delete
```
        .. code-block:: python
            dde.config.set_hvd_opt_options()
```

Comment (pescap) [reply deepxde/optimizers/config.py:87]: Sorry I did not check the doc output.

File: deepxde/optimizers/__init__.py (hunk 7 above)
Comment (pescap) [reply deepxde/optimizers/__init__.py]: > > Well, it works fine without the use of `global`.
> 
> Really? Without `global`, the code can definitely run, but you can also change `hvd_opt_options`?

Yes. It sounds like the entries of the dictionary can be edited without using global.

File: deepxde/optimizers/config.py [hunk 11]
```diff
@@ -1,9 +1,11 @@
 __all__ = ["set_LBFGS_options"]
```
Comment (pescap) [inline deepxde/optimizers/config.py:1]: Add `set_hvd_opt_options` ?

**Comment by pescap** [comment]
> > Well, it works fine without the use of `global`.
> 
> Really? Without `global`, the code can definitely run, but you can also change `hvd_opt_options`?

According to my tests, we could also remove: `global LBFGS_options`

File: deepxde/optimizers/config.py [hunk 12]
```diff
@@ -59,7 +61,40 @@ def set_LBFGS_options(
...
+
+    The default parameters are the same as for `hvd.DistributedOptimizer <https://horovod.readthedocs.io/en/stable/api.html>`_.
+
+    Args:
+        compression: Compression algorithm used to reduce the amount of data
+            sent and received by each worker node.  Defaults to not using compression.
+        op: The reduction operation to use when combining gradients across different ranks. Defaults to Average.
+        backward_passes_per_step (int): Number of backward passes to perform before calling
+            hvd.allreduce. This allows accumulating updates over multiple mini-batches before reducing and applying them.
+        average_aggregated_gradients (bool): Whether to average the aggregated gradients that have been accumulated over multiple mini-batches. If true divides gradient updates by backward_passes_per_step. Only applicable for backward_passes_per_step > 1.
```
Comment (lululxvi) [inline deepxde/optimizers/config.py:80]: This line is too long. Use 88 characters max per line

**Pull request merged**